### PR TITLE
Add utility class Defer

### PR DESCRIPTION
**Pull request by cheng-chang**
Summary:
Add a utility class `Defer` to defer the execution of a function until the Defer object goes out of scope.
Used in VersionSet:: ProcessManifestWrites as an example.
The inline comments for class `Defer` have more details.

Test Plan:
`make defer_test version_set_test && ./defer_test && ./version_set_test`

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -595,6 +595,7 @@ TESTS = \
 	db_secondary_test \
 	block_cache_tracer_test \
 	block_cache_trace_analyzer_test \
+	defer_test \
```
Comment (riversand963) [inline Makefile]: Nit: you also need to update src.mk and  CMakelists.txt. Then run `python buckifier/buckify_rocksdb.py` in order to update TARGETS file.

File: util/defer.h [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// The above code ensures that cleanups will always happen on returning.
+//
+// Without the help of Defer, you can
+// 1. every time when !s.ok(), do the cleanup;
+// 2. instead of returning when !s.ok(), continue the work only when s.ok(),
+//    but sometimes, this might lead to nested blocks of "if (s.ok()) {...}".
+//
+// With the help of Defer, you can centralize the cleanup logic inside the
+// lambda passed to Defer, and you can return immediately on failure when necessary.
+class Defer {
```
Comment (riversand963) [inline util/defer.h]: Nit: do you intend  this class to be inherited? If not, can we make it final? Do you see any potential issue with this? Just a thought.

Comment (cheng-chang) [reply util/defer.h]: Yes, I think this should be final.

File: Makefile (hunk 1 above)
Comment (cheng-chang) [reply Makefile]: updated, thanks.

**Comment by cheng-chang** [comment]
The failed test is the flaky test in env_test.

**Pull request closed without merging**